### ISSUE TITLE
`let!` doesn't infer return types

### ISSUE THREAD

**Issue by andriusch** [icy-arctic-fox/spectator#4]
```crystal
require "spectator"

def four
  2 + 2
end

Spectator.describe Int32 do
  let!(:x) { four }
  it("returns 4") { expect(x).to eq(4) }
end
```

Does not compile because crystal can't infer what type `four` returns. Strangely though if I replace `let!` with `let` it works fine and passes as it should. While I get why it doesn't compile with naive implementation of `let!`, you seem to have solved it with `let`, so it would be nice if it worked with `let!` too.

**Comment by icy-arctic-fox** [comment]
Yeah, sadly this fails because Crystal doesn't allow the following:

```crystal
def four
  2 + 2
end

class Foo
  @var = four
end
```

There's actually a `TODO` in the code where this is a problem. I'll see what I can do to work around it.

**Comment by icy-arctic-fox** [comment]
Didn't mean to close this, but can you confirm that this is fixed in the latest version on master (0.9.2)? The code in your example works now. Hopefully that fixes any other issues with it that are non-trivial.

**Comment by andriusch** [comment]
Yup it works now, very interesting way to bypass restrictions of crystal :slightly_smiling_face:
